### PR TITLE
AK/MIN-446. Chainlink price manager.

### PR DESCRIPTION
**Pull request by kurotych**
**Description:**

This is the basic implementation of Chainlink Price Manager.
For test staging and proof of concept.

Approach in this code can be reviewed and changed in future.

In next versions, TODOs will be fixed and should be added two providers of feeds: Chainlink and Minterest

**Changes in storage or API:**

There are no changes in API.

**Checklist**

- [x] Code is covered with tests
- [x] Code is covered with benchmarks
- [ ] Fresh benchmark results included

### CODE REVIEW

File: pallets/chainlink-price-manager/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+		// TODO handle updated_at
+		let RoundData { answer, .. } = feed_result.latest_data();
+
+		// There is an issue that pallet-chainlink-feed can return Some(0)
+		// if feed was created but submit() extrinsic wasn't called
+		if answer.is_zero() {
+			return None;
+		}
+
+		Some(Price::from_inner(answer.into()))
```
Comment (lut1k) [inline pallets/chainlink-price-manager/src/lib.rs:250]: The associated type Value implement `trait BaseArithmetic`: https://github.com/smartcontractkit/chainlink-polkadot/blob/master/pallet-chainlink-feed/src/lib.rs#L298.
`traits BaseArithmetic` depends on traits `TryFrom<u128> + TryInto<u128> + UniqueSaturatedFrom<u128> + UniqueSaturatedInto<u128>`: https://substrate.dev/rustdocs/v3.0.0-monthly-2021-05/sp_arithmetic/traits/trait.BaseArithmetic.html
So I think there is no need to add a constraint:
`where  u128: From<<T as pallet_chainlink_feed::Config>::Value>,`
and you can write like this:
 ```suggestion
		Some(Price::from_inner(answer.try_into().ok()?))
```
or
```suggestion
		Some(Price::from_inner(answer.unique_saturated_into()))
```

File: runtime/src/lib.rs [hunk 2]
```diff
@@ -578,6 +582,38 @@ impl whitelist_module::Config for Runtime {
...
 
+impl chainlink_price_manager::Config for Runtime {
+	type Event = Event;
+	type PalletAccountId = ChainlinkPriceManagerAccountId;
+	type UnsignedPriority = ChainlinkManagerPriority;
+	type ChainlinkPriceManagerWeightInfo = weights::chainlink_price_manager::WeightInfo<Runtime>;
+}
+
+pub type FeedId = u32;
+pub type Value = u128;
```
Comment (lut1k) [inline runtime/src/lib.rs:593]: I think this types should be moved to he primitives crate.

File: runtime/src/lib.rs [hunk 3]
```diff
@@ -578,6 +582,38 @@ impl whitelist_module::Config for Runtime {
...
+	pub const StringLimit: u32 = 30;
+	pub const OracleCountLimit: u32 = 25;
+	pub const FeedLimit: FeedId = 100;
+	// TODO Review this parameter before production preperements
+	pub const MinimumReserve: Balance = ExistentialDeposit::get() * 1000;
+}
+
+impl pallet_chainlink_feed::Config for Runtime {
+	type Event = Event;
+	type FeedId = u32;
```
Comment (lut1k) [inline runtime/src/lib.rs:604]:
```suggestion
	type FeedId = FeedId;
```

File: service/src/chain_spec.rs [hunk 4]
```diff
@@ -464,6 +465,10 @@ fn minterest_genesis(
 			members: whitelist_members,
 			whitelist_mode: false,
 		},
+		chainlink_feed: ChainlinkFeedConfig {
+			pallet_admin: Some(root_key),
+			feed_creators: endowed_accounts,
```
Comment (lut1k) [inline service/src/chain_spec.rs:470]: Why are the `endowed_accounts` in the `minterest_genesis`, and the `root_key` in the `testnet_genesis`?
`endowed_accounts` - accounts to which the existential balance should be deposited. There are not the same accounts as `feed_creators`. So, suggestion:
```suggestion
			feed_creators: vec![root_key],
```

File: types.json [hunk 5]
```diff
@@ -192,7 +192,10 @@
         "reserved": "Balance",
         "frozen": "Balance"
       },
-      "AccountData": "OrmlAccountData"
+      "AccountData": "OrmlAccountData",
+      "FeedId": "u32",
+      "SubmissionBounds": "(u32, u32)",
+      "RoundId": "u32"
```
Comment (lut1k) [inline types.json:198]: Shouldn't we add the type `"Value": "u128"` ?

File: runtime/src/lib.rs [hunk 6]
```diff
@@ -578,6 +582,38 @@ impl whitelist_module::Config for Runtime {
...
+	type Value = Value;
```
Comment (lut1k) [inline runtime/src/lib.rs:605]:
```suggestion
	type FeedId = FeedId;
	type Value = Balance;
```

File: pallets/chainlink-price-manager/src/lib.rs (hunk 1 above)
Comment (kurotych) [reply pallets/chainlink-price-manager/src/lib.rs]: Nice suggestion :+1:  Improved

File: runtime/src/lib.rs (hunk 2 above)
Comment (kurotych) [reply runtime/src/lib.rs]: Moved

File: service/src/chain_spec.rs (hunk 4 above)
Comment (kurotych) [reply service/src/chain_spec.rs]: Fixed. Thanks

File: types.json (hunk 5 above)
Comment (lut1k) [inline types.json:198]:
```
"ChainlinkFeedId": "u32",
"ChainlinkPriceValue": "u128",
```

File: pallets/chainlink-price-manager/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+					"Initiate a new round is faled",
+				);
+			}
+		}
+		Self::print_prices();
+		Ok(())
+	}
+
+	// TODO This is temporary function. We need tom move this function as method to
+	// privitives/src/currency.rs. Also, add distingiush between chainlink provider and minterest
```
Comment (ales-by) [inline pallets/chainlink-price-manager/src/lib.rs]:
```suggestion
	// primitives/src/currency.rs. Also, add distingiush between chainlink provider and minterest
```

**Review by ales-by — DISMISSED** [review]

File: types.json (hunk 5 above)
Comment (kurotych) [reply types.json]: Added

File: pallets/chainlink-price-manager/src/lib.rs (hunk 7 above)
Comment (kurotych) [reply pallets/chainlink-price-manager/src/lib.rs]: Fixed

**Review by lut1k — APPROVED** [review]

**Review by Apollo50 — APPROVED** [review]

**Pull request merged**